java.lang.AssertionError in compiler with SAM traits with given parameters

## minimized code

```Scala
trait Evidence[X]

trait Trait[X : Evidence]
  def method(x : X) : X

given ev : Evidence[Int] = new Evidence[Int]{}
val crash : Trait[Int] = (x: Int) => x
```

<details>
<summary>Compilation output</summary>

```scala
java.lang.AssertionError: assertion failed: missing parameters for trait Trait from () extends Object(), Trait {
  final def method(x: Int): Int = $anonfun(x)
  final def method(x: Object): Object =
    scala.Int.box(this.method(scala.Int.unbox(x)))
} should have been caught in typer while compiling <my scala file>
```
</details>

## expectation
I think either:
- This should give a proper compiler error, forcing the user to write
```Scala
val noCrash : Trait[Int] = new Trait[Int] with 
  def method(x: Int) = x
```
which does compile succesfully.
- This should work if the evidence is in scope, altough this would force you to use summon
- The SAM functionality could be adapted to support given parameters in the trait. In this case I think it would make a lot of sense if the following example would work:
```Scala
val noCrash : Trait[Int] = (given Evidence[Int]) => (x: Int) => x
```

--- a/compiler/src/dotty/tools/dotc/core/Types.scala
+++ b/compiler/src/dotty/tools/dotc/core/Types.scala
@@ -4518,7 +4518,7 @@ object Types {
           case et: ExprType => true
           case _ => false
         }
-        if (tp.cls.is(Trait) || zeroParams(tp.cls.primaryConstructor.info)) tp // !!! needs to be adapted once traits have parameters
+        if (zeroParams(tp.cls.primaryConstructor.info)) tp
         else NoType
       case tp: AppliedType =>
         zeroParamClass(tp.superType)

--- a/compiler/src/dotty/tools/dotc/core/Types.scala
+++ b/compiler/src/dotty/tools/dotc/core/Types.scala
@@ -4518,7 +4518,9 @@ object Types {
           case et: ExprType => true
           case _ => false
         }
-        if (zeroParams(tp.cls.primaryConstructor.info)) tp
+        // `ImplicitFunctionN` does not have constructors
+        val ctor = tp.cls.primaryConstructor
+        if (!ctor.exists || zeroParams(ctor.info)) tp
         else NoType
       case tp: AppliedType =>
         zeroParamClass(tp.superType)

--- a/tests/neg/i7980.scala
+++ b/tests/neg/i7980.scala
@@ -1,0 +1,7 @@
+trait Evidence[X]
+
+trait Trait[X : Evidence]
+  def method(x : X) : X
+
+given ev : Evidence[Int] = new Evidence[Int]{}
+val crash : Trait[Int] = (x: Int) => x          // error

